File: src/TemporalProlog/Batch.hs
-- |
-- Module      : TemporalProlog.Batch
-- Description : Deterministic branch-preserving batch execution
--
-- Provides the non-interactive execution contract shared with the Rust
-- implementation.  Every minimal-model branch is retained, sorted
-- canonically, and rendered with its complete observable history.
module TemporalProlog.Batch
  ( BatchOptions(..)
  , BatchResult(..)
  , runBatch
  , runBatchWithAuxiliaries
  , renderBatch
  , semanticDigest
  ) where

import Control.Monad (foldM, forM, forM_, unless, when)
import Data.Bits (xor)
import qualified Data.ByteString as ByteString
import Data.List (intercalate, sortOn)
import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map)
import qualified Data.Set as Set
import Data.Set (Set)
import qualified Data.Text as Text
import qualified Data.Text.Encoding as TextEncoding
import Data.Word (Word8, Word64)
import Numeric (showHex)

import TemporalProlog.Interpreter
import TemporalProlog.Syntax

-- | Finite execution horizon, zero-indexed external inputs, and rendering
-- visibility for one batch run.
data BatchOptions = BatchOptions
  { batchSteps           :: Int             -- ^ Positive number of worlds.
  , batchAssertions      :: Map Int [Atom]  -- ^ Ground facts by world number.
  , batchIncludeInternal :: Bool            -- ^ Display generated auxiliaries.
  } deriving (Eq, Show)

-- | Completed branch-preserving execution and the options that produced it.
data BatchResult = BatchResult
  { batchResultOptions  :: BatchOptions
  , batchResultBranches :: [InterpreterState]
  , batchResultAuxiliaryPredicates :: Set Name
  } deriving (Show)

-- | Execute a finite schedule while preserving every minimal-model branch.
-- Invalid or unreachable scheduled inputs are rejected instead of silently
-- being ignored.
runBatch
  :: BatchOptions
  -> NormalProgram
  -> Set Name
  -> Either String BatchResult
runBatch options program pfNames =
  runBatchWithAuxiliaries options program pfNames Set.empty

-- | Batch execution with exact normalizer-generated predicate metadata for
-- user-facing rendering.
runBatchWithAuxiliaries
  :: BatchOptions
  -> NormalProgram
  -> Set Name
  -> Set Name
  -> Either String BatchResult
runBatchWithAuxiliaries options program pfNames auxiliaryPredicates = do
  validateOptions options
  branches <- go 0
    [newInterpreterStateWithAuxiliaries program pfNames auxiliaryPredicates]
  Right BatchResult
    { batchResultOptions = options
    , batchResultBranches = sortOn branchKey branches
    , batchResultAuxiliaryPredicates = auxiliaryPredicates
    }
  where
    go step branches
      | step >= batchSteps options = Right branches
      | otherwise = do
          next <- fmap concat $ forM branches $ \branch -> do
            asserted <- foldM (flip assertFactEither) branch
              (Map.findWithDefault [] step (batchAssertions options))
            stepWorldAll asserted
          go (step + 1) (sortOn branchKey next)

    branchKey = map worldToSet . getHistory

validateOptions :: BatchOptions -> Either String ()
validateOptions options = do
  when (batchSteps options <= 0) $
    Left "steps must be a positive integer"
  forM_ (Map.toList (batchAssertions options)) $ \(step, atoms) -> do
    when (step < 0) $ Left "assertion steps must be non-negative"
    when (step >= batchSteps options) $ Left $
      "assertion step " ++ show step ++ " is outside the "
      ++ show (batchSteps options) ++ "-world run"
    unless (all isGroundAtom atoms) $
      Left "assertions must be ground"

-- | Render a stable, line-oriented result.  The digest covers complete raw
-- worlds, including generated auxiliaries, while the fact lists hide them by
-- default for a user-facing view.
renderBatch :: BatchResult -> String
renderBatch result = unlines $
  [ "steps=" ++ show (batchSteps options)
  , "branches=" ++ show (length branches)
  ] ++ concat
    [ renderBranch index branch
    | (index, branch) <- zip [0 :: Int ..] branches
    ]
  where
    options = batchResultOptions result
    branches = batchResultBranches result
    auxiliaryPredicates = batchResultAuxiliaryPredicates result

    renderBranch index branch =
      ("branch=" ++ show index) :
      [ "  w" ++ show worldNumber ++ "=" ++ renderWorld world
      | (worldNumber, world) <- zip [0 :: Int ..] (getHistory branch)
      ] ++ ["  digest=" ++ semanticDigest (getHistory branch)]

    renderWorld world = "[" ++ intercalate "," visible ++ "]"
      where
        visible =
          [ canonicalAtom atom
          | atom <- Set.toAscList (worldToSet world)
          , batchIncludeInternal options
              || not (internalAtom auxiliaryPredicates atom)
          ]

-- | FNV-1a checksum of a complete history using the same canonical byte
-- stream as the Rust engine.
semanticDigest :: [World] -> String
semanticDigest worlds = pad16 $ showHex digest ""
  where
    bytes = TextEncoding.encodeUtf8 . Text.pack $ concat
      [ show index ++ ":" ++ concatMap ((++ ";") . canonicalAtom)
          (Set.toAscList (worldToSet world))
      | (index, world) <- zip [0 :: Int ..] worlds
      ]
    digest = ByteString.foldl' step 14695981039346656037 bytes
    step :: Word64 -> Word8 -> Word64
    step hash byte = (hash `xor` fromIntegral byte) * 1099511628211
    pad16 value = replicate (16 - length value) '0' ++ value

canonicalAtom :: Atom -> String
canonicalAtom (Atom name []) = name
canonicalAtom (Atom name terms) =
  name ++ "(" ++ intercalate "," (map canonicalTerm terms) ++ ")"

canonicalTerm :: Term -> String
canonicalTerm (TVar variable) = variable
canonicalTerm (TFun name []) = name
canonicalTerm (TFun name terms) =
  name ++ "(" ++ intercalate "," (map canonicalTerm terms) ++ ")"
canonicalTerm (TPrev term) = "@" ++ canonicalTerm term

internalAtom :: Set Name -> Atom -> Bool
internalAtom _ (Atom "true" []) = True
internalAtom _ (Atom "at" _) = True
internalAtom auxiliaryPredicates (Atom name _) =
  name `Set.member` auxiliaryPredicates

File: rust/src/batch.rs
use crate::{semantic_digest, Atom, Interpreter, NormalizedProgram, Term};
use std::collections::BTreeMap;

/// Finite execution horizon, scheduled inputs, and rendering visibility.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BatchOptions {
    /// Positive number of worlds to compute.
    pub steps: usize,
    /// Ground external facts indexed by zero-based world number.
    pub assertions: BTreeMap<usize, Vec<Atom>>,
    /// Whether rendered histories expose normalization auxiliaries.
    pub include_internal: bool,
}

impl Default for BatchOptions {
    fn default() -> Self {
        Self {
            steps: 1,
            assertions: BTreeMap::new(),
            include_internal: false,
        }
    }
}

/// Completed branch-preserving execution and its options.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BatchResult {
    pub options: BatchOptions,
    pub branches: Vec<Interpreter>,
}

/// Execute a finite schedule while preserving every minimal-model branch.
/// Inputs outside the selected horizon are errors rather than ignored data.
pub fn run_batch(options: BatchOptions, program: NormalizedProgram) -> Result<BatchResult, String> {
    validate_options(&options)?;
    let mut branches = vec![Interpreter::new(program)];
    for step in 0..options.steps {
        let mut next = Vec::new();
        for mut branch in branches {
            for atom in options.assertions.get(&step).into_iter().flatten() {
                branch.assert(atom.clone())?;
            }
            next.extend(branch.step_all()?);
        }
        next.sort_by(|left, right| left.worlds.cmp(&right.worlds));
        branches = next;
    }
    branches.sort_by(|left, right| left.worlds.cmp(&right.worlds));
    Ok(BatchResult { options, branches })
}

fn validate_options(options: &BatchOptions) -> Result<(), String> {
    if options.steps == 0 {
        return Err("steps must be a positive integer".into());
    }
    for (step, atoms) in &options.assertions {
        if *step >= options.steps {
            return Err(format!(
                "assertion step {step} is outside the {}-world run",
                options.steps
            ));
        }
        if atoms.iter().any(|atom| !atom.ground()) {
            return Err("assertions must be ground".into());
        }
    }
    Ok(())
}

/// Render the common deterministic, line-oriented batch format.
pub fn render_batch(result: &BatchResult) -> String {
    let mut output = format!(
        "steps={}\nbranches={}\n",
        result.options.steps,
        result.branches.len()
    );
    for (index, branch) in result.branches.iter().enumerate() {
        output.push_str(&format!("branch={index}\n"));
        for (world_number, world) in branch.worlds.iter().enumerate() {
            let facts = world
                .iter()
                .filter(|atom| {
                    result.options.include_internal
                        || !internal_atom(atom, &branch.program.auxiliary_predicates)
                })
                .map(canonical_atom)
                .collect::<Vec<_>>()
                .join(",");
            output.push_str(&format!("  w{world_number}=[{facts}]\n"));
        }
        output.push_str(&format!("  digest={}\n", semantic_digest(&branch.worlds)));
    }
    output
}

fn canonical_atom(atom: &Atom) -> String {
    if atom.terms.is_empty() {
        atom.name.clone()
    } else {
        format!(
            "{}({})",
            atom.name,
            atom.terms
                .iter()
                .map(canonical_term)
                .collect::<Vec<_>>()
                .join(",")
        )
    }
}

fn canonical_term(term: &Term) -> String {
    match term {
        Term::Var(variable) => variable.clone(),
        Term::Fun(name, terms) if terms.is_empty() => name.clone(),
        Term::Fun(name, terms) => format!(
            "{}({})",
            name,
            terms
                .iter()
                .map(canonical_term)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Term::Prev(term) => format!("@{}", canonical_term(term)),
    }
}

fn internal_atom(atom: &Atom, auxiliary_predicates: &std::collections::BTreeSet<String>) -> bool {
    atom.name == "true" || atom.name == "at" || auxiliary_predicates.contains(&atom.name)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{compile, parse_atom};

    #[test]
    fn preserves_and_canonically_renders_all_branches() {
        let result = run_batch(
            BatchOptions::default(),
            compile("~a => b. ~b => a.").unwrap(),
        )
        .unwrap();
        assert_eq!(result.branches.len(), 2);
        assert_eq!(
            render_batch(&result),
            "steps=1\nbranches=2\n\
             branch=0\n  w0=[a]\n  digest=3ac3fa0d287999b7\n\
             branch=1\n  w0=[b]\n  digest=faa2015c6af3d282\n"
        );
    }

    #[test]
    fn rejects_unreachable_and_nonground_assertions() {
        let program = compile("ok.").unwrap();
        let mut unreachable = BatchOptions::default();
        unreachable
            .assertions
            .insert(1, vec![parse_atom("event").unwrap()]);
        assert!(run_batch(unreachable, program.clone()).is_err());

        let mut nonground = BatchOptions::default();
        nonground
            .assertions
            .insert(0, vec![parse_atom("event(X)").unwrap()]);
        assert!(run_batch(nonground, program).is_err());
    }

    #[test]
    fn source_aux_suffixes_remain_visible() {
        let mut options = BatchOptions {
            steps: 2,
            ..BatchOptions::default()
        };
        options
            .assertions
            .insert(0, vec![parse_atom("trigger").unwrap()]);
        let result = run_batch(
            options,
            compile("user_aux0. always_aux0. trigger => next generated.").unwrap(),
        )
        .unwrap();
        let rendered = render_batch(&result);
        assert!(rendered.contains("w0=[always_aux0,trigger,user_aux0]"));
        assert!(!rendered.contains("next_aux1"));
    }
}
